src/BasicRum/Diagram/Builder/RenderType/Plane.php

```diff
@@ -0,0 +1,103 @@
+<?php
+
+declare(strict_types=1);
+
+namespace App\BasicRum\Diagram\Builder\RenderType;
+
+use App\BasicRum\Diagram\Builder\RenderType\Metrics\PlaneBusinessMetrics;
+use App\BasicRum\Diagram\Builder\RenderType\Metrics\PlaneTechnicalMetrics;
+use App\BasicRum\Diagram\View\Layout;
+use App\BasicRum\Diagram\View\RenderType\Plane as ViewRenderTypePlane;
+use App\BasicRum\DiagramOrchestrator;
+use App\BasicRum\Release;
+use App\BasicRum\RenderTypeInterface;
+
+class Plane implements RenderTypeInterface
+{
+    private $diagramOrchestrator;
+    private $params;
+    private $releaseRepository;
+    private $dataForDiagram;
+    private $extraDiagramParams;
+    private $extraLayoutParams;
+    private $results;
+
+    public function __construct(DiagramOrchestrator $diagramOrchestrator, array $params, Release $releaseRepository)
+    {
+        $this->diagramOrchestrator = $diagramOrchestrator;
+        $this->params = $params;
+        $this->releaseRepository = $releaseRepository;
+        $this->dataForDiagram = [];
+        $this->extraDiagramParams = [];
+        $this->extraLayoutParams = $this->getExtraLayoutParams($params);
+        $this->results = $diagramOrchestrator->process();
+    }
+
+    public function build(): array
+    {
+        $hasError = false;
+
+        try {
+            $businessMetrics = new PlaneBusinessMetrics($this->results, $this->params);
+            $technicalMetrics = new PlaneTechnicalMetrics($this->results, $this->params);
+            foreach ($this->results as $key => $result) {
+                $businessMetrics->proceed($key);
+                $technicalMetrics->proceed($key);
+            }
+
+            $this->setProperty($this->dataForDiagram, $businessMetrics->getDataForDiagram());
+            $this->setProperty($this->extraDiagramParams, $businessMetrics->getExtraDiagramParams());
+            $this->setProperty($this->extraLayoutParams, $businessMetrics->getExtraLayoutParams());
+
+            $this->setProperty($this->dataForDiagram, $technicalMetrics->getDataForDiagram());
+            $this->setProperty($this->extraDiagramParams, $technicalMetrics->getExtraDiagramParams());
+            $this->setProperty($this->extraLayoutParams, $technicalMetrics->getExtraLayoutParams());
+        } catch (\Throwable $e) {
+            $hasError = true;
+        }
+
+        $view = new ViewRenderTypePlane(new Layout());
+
+        // It has to be sorted. Because order here has huge impact
+        // probably object would be more suitable here
+        $this->sortDataForDiagram();
+
+        return $view->build(
+            $this->dataForDiagram,
+            $this->params,
+            $this->extraLayoutParams,
+            $this->extraDiagramParams,
+            $hasError
+        );
+    }
+
+    private function sortDataForDiagram(): void
+    {
+        ksort($this->dataForDiagram);
+    }
+
+    private function setProperty(&$property, ?array $data): void
```

Comment:
What is the idea of this method? Why don't we assign directly values to:
 -  $this->dataForDiagram
 - $this->extraDiagramParams
and so on ...

Reply:
This method allows to preserve keys order. Otherwise array keys losing their sync with $params array keys and later in BasicRum/Diagram/View/RenderType/Plane.php  in build function on lines: 55, 57, 59 this lead to error because it is looking for $key from $samples (which is actually $this->dataForDiagram) in $renderParams (which is actually $params) and can not find it there or getting something else than expected.

Reply:
Hello @korzol 

I am ready to merge the PR. I have only one question:

Are the test failing when the keys are not in order?

Reply:
Hello @ceckoslab 
Yes, the following tests are fails when keys are not in correct order
testBounceRateValuesInBuckets
testBounceRateAndFirstByteCorrectDiagramNames

These tests are resides in FirstByteAndBounceRateTest.php